sdk/core/Azure.Core.Experimental/src/Auth/OAuth2BearerTokenAuthenticationPolicy.cs

```diff
@@ -0,0 +1,76 @@
+/// </summary>
+/// <remarks>
+/// Initializes a new instance of the <see cref="OAuth2BearerTokenAuthenticationPolicy"/> class.
+/// </remarks>
+public class OAuth2BearerTokenAuthenticationPolicy : PipelinePolicy
+{
+    private readonly ITokenProvider _tokenProvider;
+    private readonly IScopedFlowToken _flowContext;
+
+    /// <param name="tokenProvider"></param>
+    /// <param name="contexts"></param>
+    public OAuth2BearerTokenAuthenticationPolicy(ITokenProvider tokenProvider, IEnumerable<IReadOnlyDictionary<string, object>> contexts)
+    {
+        _tokenProvider = tokenProvider;
+        _flowContext = GetContext(contexts, tokenProvider);
+    }
+
+    /// <inheritdoc />
+    public override void Process(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex)
+    {
+        Token token;
+        if (message.TryGetProperty(typeof(IScopedFlowToken), out var rawContext) && rawContext is IScopedFlowToken scopesContext)
+        {
+            var context = _flowContext.CloneWithAdditionalScopes(scopesContext.Scopes);
+            token = _tokenProvider.GetAccessToken(context, message.CancellationToken);
+        }
+        else
+        {
+            token = _tokenProvider.GetAccessToken(_flowContext, message.CancellationToken);
+        }
+        message.Request.Headers.Set("Authorization", $"Bearer {token.TokenValue}");
+
+        ProcessNext(message, pipeline, currentIndex);
+        return;
+    }
+
+    /// <inheritdoc />
+    public override ValueTask ProcessAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex)
+    {
+        return ProcessNextAsync(message, pipeline, currentIndex);
+    }
+
+    internal static IScopedFlowToken GetContext(IEnumerable<IReadOnlyDictionary<string, object>> contexts, ITokenProvider tokenProvider)
+    {
+        var type = tokenProvider.GetType();
+        // This assumes that a credential provider will only implement one flow type.
+        var credentialFlowType = type switch
+        {
+            var t when typeof(TokenProvider<IClientCredentialsFlowToken>).IsAssignableFrom(t) => typeof(IClientCredentialsFlowToken),
+            var t when typeof(TokenProvider<IAuthorizationCodeFlowToken>).IsAssignableFrom(t) => typeof(IAuthorizationCodeFlowToken),
+            var t when typeof(TokenProvider<IPasswordFlowToken>).IsAssignableFrom(t) => typeof(IPasswordFlowToken),
+            var t when typeof(TokenProvider<IImplicitFlowToken>).IsAssignableFrom(t) => typeof(IImplicitFlowToken),
+            _ => throw new InvalidOperationException("Supplied credential does not implement any supported auth flow.")
+        };
+        foreach (var context in contexts)
+        {
+            var createdContext = tokenProvider.CreateContext(context);
+            if (createdContext != null && credentialFlowType.IsInstanceOfType(createdContext))
+            {
+                return (IScopedFlowToken)createdContext;
+            }
```

Comment:
In theory this could be done without the type inspection above and the `credentialFlowType.IsInstanceOfType(createdContext)` check as long as the provider is particular enough with it's validation in its `CreateContext` implementation.

Reply:
Is that approach AOT-friendly?   I think it should be, but just to be sure - @m-redding?

Reply:
It seems fine to me! But it would be good to run the script to double check - `./eng/scripts/compatibility/Check-AOT-Compatibility.ps1 Azure.Core.Experimental` from root. But, for this one since the namespace and the package name aren't the same, the namespace would need to be hard coded. [L40 ](https://github.com/Azure/azure-sdk-for-net/blob/main/eng/scripts/compatibility/Check-AOT-Compatibility.ps1#L40) can be hard coded to `using System.ClientModel;` instead of `using $PackageName;` 
I can make a fix to the script and add an option to pass a namespace name. Although, looking at the latest AOT docs it seems like the using statement isn't necessary anymore to get the warnings.